PSA: TFM import

urutva opened the pull request:
### Description

Import TF-M sources from master branch at hash e7efdc6e8d89a22c4cac5b6ccc8e7d0d332ff034


### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@alzix @orenc17 @Patater 
### Release Notes

Commit:
TF-M patch: General modifications

- Remove un-needed files
- Disable printf and uart
- Modify include paths
- Guard macros from mbed_lib with ifndef

(cherry picked from commit 1f30b52488e88704cdb51c8c26e7225e580e5cc2)
(cherry picked from commit 71cd34df3265d98da5c9b34e4e18ef039d7bef5c)
(cherry picked from commit 185d2865da45cc2c6ac3acb755b90c196934d7d5)
(cherry picked from commit fb068d2cb4e89cacf0e9f413075bb4b211f1484f)

Commit:
TF-M patch: Fix tfm_ns_lock_init issue (TF-M issue #239)

- Link to bug tracking: https://developer.trustedfirmware.org/T239

(cherry picked from commit 5f2e4b3911ca3a776fab55a6d12054904614cced)
(cherry picked from commit 5d41a2aeae71b13f2763bea4e55899646291e0eb)

Commit:
TF-M patch: Fix service handles not cleared issue (TF-M issue #230)

- Link to bug tracking: https://developer.trustedfirmware.org/T230

(cherry picked from commit 0c23e8698958b6e716114267fbdf6d82a16b6e0c)
(cherry picked from commit 9c1e080e39adc7211c8c2c12cd652da3dc124299)

Commit:
TF-M patch: Fix tfm_psa_call_venner wrong argument type (TF-M issue #241)

- Link to bug tracking: https://developer.trustedfirmware.org/T241

(cherry picked from commit da01e3411fcf8010f4ae581946cf4b0e7753e024)
(cherry picked from commit 280715f9b74ab29459d81edaf02b39e7a6acb13c)

urutva committed:
[trusted-firmware-m]: Updated to e7efdc6

urutva committed:
TF-M patch: Change #if TFM_PSA_API to #ifdef TFM_PSA_API to avoid
compiler errors as mbed-cli only generates "-D" macros only for
"macros" defined in targets.json

TF-M task link: https://developer.trustedfirmware.org/T396

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
TF-M patch: Remove secure_fw/core/tfm_func_api.c which is required only
when TFM_PSA_API is not set

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/include/tfm_spm.h:
```diff
@@ -279,22 +289,38 @@ int32_t tfm_spm_check_client_version(struct tfm_spm_service_t *service,
+ *  ctxb        -    State context storage pointer
+ *
+ * Notes:
+ *  This is a staging API. Scheduler should be called in SPM finally and
```
In light of this comment should this be marked as deprecated in some way?

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/include/tfm_thread.h:
```diff
@@ -201,23 +214,19 @@ void tfm_thrd_context_switch(struct tfm_state_context_ext *ctxb,
+ *  Must be called in thread mode.
  */
-void tfm_pendsv_do_schedule(struct tfm_state_context_ext *ctxb);
+void tfm_thrd_exit(void);
```
Why abbreviate so aggressively? tfm_thread_exit(), anyone?

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/tfm_spm.c:
```diff
@@ -481,25 +468,37 @@ int32_t tfm_memory_check(void *buffer, size_t len, int32_t ns_caller,
     return IPC_ERROR_MEMORY_CHECK;
 }
 
+uint32_t tfm_spm_partition_get_privileged_mode(uint32_t partition_idx)
```
Personally I'd prefer something along the lines of '..._is_priviledged_mode()' that returns TRUE or FALSE, but that might be too many years of Java. The function as it stands will still need its (typeless) return value checking against one of these constants; an opportunity for a bug later...

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/tfm_spm.c:
```diff
@@ -559,6 +561,58 @@ void tfm_spm_init(void)
+                                                 struct spm_partition_desc_t,
+                                                 sp_thrd);
+
+        if (p_next_partition->static_data.partition_flags &
```
Looks like a duplicate of that 'tfm_spm_partition_get_privileged_mode()' function above.

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/tfm_thread.c:
```diff
@@ -124,75 +126,56 @@ void tfm_thrd_set_status(struct tfm_thrd_ctx *pth, uint32_t new_status)
-        tfm_thrd_context_switch(ctxb, CURR_THRD, pth);
-    }
+    /* Update current thread indicator */
+    CURR_THRD = next;
```
CURRENT_THREAD

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/tfm_thread.c:
```diff
@@ -124,75 +126,56 @@ void tfm_thrd_set_status(struct tfm_thrd_ctx *pth, uint32_t new_status)
 
 /* Remove current thread out of the schedulable list */
-void tfm_thrd_do_exit(void)
+void tfm_svcall_thrd_exit(void)
```
thrd->thread. This is currently inconsistent with other functions defined, eg 'tfm_nspm_thread_entry'

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/ipc/tfm_thread.c:
```diff
@@ -124,75 +126,56 @@ void tfm_thrd_set_status(struct tfm_thrd_ctx *pth, uint32_t new_status)
 
-/* Scheduling won't happen immediately but after the exception returns */
-void tfm_thrd_activate_schedule(void)
+void tfm_thrd_start_scheduler(struct tfm_thrd_ctx *pth)
```
thrd->thread

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/core/tfm_svc.h:
```diff
@@ -23,6 +23,7 @@ typedef enum {
 #ifdef TFM_PSA_API
     TFM_SVC_IPC_REQUEST,
     TFM_SVC_SCHEDULE,
+    TFM_SVC_EXIT_THRD,
```
TFM_SVC_EXIT_THREAD

mark-edgeworth commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/spm/spm_api.c:
```diff
@@ -390,3 +400,19 @@ void tfm_spm_partition_cleanup_context(uint32_t partition_idx)
+
+    ctrl.w = __get_CONTROL();
+
+    if (privileged == TFM_PARTITION_PRIVILEGED_MODE) {
```
I'd prefer a ternary operator here.

mark-edgeworth commented on tools/importer/tfm_importer.json:
```diff
@@ -64,12 +68,12 @@
-    "280715f9b74ab29459d81edaf02b39e7a6acb13c",
-    "ea81bf91c90ae23dd9de012bfd7498613be00601",
-    "5342015bb12a486a1c563175a8a7129f0737c925"
+    "11bff3f3cbfbd3e2c284e884d0066531e6b47d7e",
```
Don't know what these are - should they be committed?

Patater replied:
These are commit hashes of patches present in the Mbed OS repo (in this very PR, in fact!), necessary to complete the import and integration of TF-M into Mbed OS. Yes, they are needed and should be committed.

mark-edgeworth reviewed (commented):
Ok, comments added as I see it. Apologies in advance if I'm out of step with the whole of Mbed OS so far, and feel free to disregard if you want. Your comments are welcome though.
Most general point - abbreviating names to this extent makes the code less readable for me.

Patater commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/spm/spm_db.h:
```diff
@@ -62,13 +62,18 @@ struct spm_partition_desc_t {
 #define REGION(a, b, c) a##b##c
 #define REGION_NAME(a, b, c) REGION(a, b, c)
-#if (TFM_LVL == 1) && !defined(TFM_PSA_API)
+#if (TFM_LVL == 1)
```
Please comment in our patch why this check is changed from the original.

urutva replied:
done

Patater commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/spm/spm_api.c:
```diff
@@ -286,7 +290,6 @@ uint32_t tfm_spm_partition_get_flags(uint32_t partition_idx)
             partition_flags;
 }
 
-#ifndef TFM_PSA_API
```
Please comment in our patch (in the code, not the commit description) why this check is removed.

urutva replied:
This change is not relevant anymore. Retained the code from TF-M.

Patater commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_SPE/secure_fw/spm/spm_db_setup.h:
```diff
@@ -38,7 +38,7 @@ struct spm_partition_db_t {
     } while (0)
 
-#if (TFM_LVL == 1) && !defined(TFM_PSA_API)
+#if (TFM_LVL == 1)
```
Please comment in our patch why this change is needed.

urutva replied:
done

Patater reviewed (changes requested):
- Do we need the patch "TF-M patch/workaround related to (TF-M issue #T240)"? TF-M has merged the fix.
- There seems to be a patch open for https://developer.trustedfirmware.org/T396, with a request for your comment. Could we use that patchset instead of our own?

alzix commented on tools/importer/tfm_importer.json:
```diff
@@ -64,12 +68,12 @@
+    "910a402ce6c96b654cb6ae1a5b679e4f856c5419",
+    "d3c610e090282893ba1820c30fc8be3c03195091",
+    "fb35880cecff7205fd19aaaed02b9958b9bfce5a",
+    "be5cfdf18e6e70c9360dea7456c85189a56f277e"
```
a while ago i've added an option to specify a help message/description for each hash. Take a look here https://github.com/ARMmbed/mbed-os/blob/master/tools/importer/importer.py#L239

Consider specifying a commit message for each hash - after couple of months when you will be back for new re-import - it will help you a lot.

urutva replied:
@alzix Ah cool. Thanks. I'll update the json file.

urutva committed:
TF-M patch/workaround related to (TF-M issue #T240)

- Link to bug tracking: https://developer.trustedfirmware.org/T240

The issue is fixed by TF-M team. However they autogenerate region details
(code, ro, rw, zi and stack ) using linker scripts and in mbed-os we
also autogenerate region details but using mix of service definition in
json file and other template files.

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

urutva committed:
Update commit sha's

Signed-off-by: Devaraj Ranganna <devaraj.ranganna@arm.com>

Patater reviewed (approved):
LGTM
